// svix/src/openapi/models/ListResponseMessageAttemptEndpointOut.ts
/**
 * Svix API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.1.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { MessageAttemptEndpointOut } from './MessageAttemptEndpointOut';
import { HttpFile } from '../http/http';

export class ListResponseMessageAttemptEndpointOut {
    'data': Array<MessageAttemptEndpointOut>;
    'done': boolean;
    'iterator'?: string;
    'prevIterator'?: string;

    static readonly discriminator: string | undefined = undefined;

    static readonly attributeTypeMap: Array<{name: string, baseName: string, type: string, format: string}> = [
        {
            "name": "data",
            "baseName": "data",
            "type": "Array<MessageAttemptEndpointOut>",
            "format": ""
        },
        {
            "name": "done",
            "baseName": "done",
            "type": "boolean",
            "format": ""
        },
        {
            "name": "iterator",
            "baseName": "iterator",
            "type": "string",
            "format": ""
        },
        {
            "name": "prevIterator",
            "baseName": "prevIterator",
            "type": "string",
            "format": ""
        }    ];

    static getAttributeTypeMap() {
        return ListResponseMessageAttemptEndpointOut.attributeTypeMap;
    }
    
    public constructor() {
    }
}
