// svix/src/openapi/models/SinkInOneOf1.ts
/**
 * Svix API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.1.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { URI } from './URI';
import { HttpFile } from '../http/http';

export class SinkInOneOf1 {
    'accessKey': string;
    'queueDsn': URI;
    'region': string;
    'secretKey': string;
    'type': SinkInOneOf1TypeEnum;

    static readonly discriminator: string | undefined = undefined;

    static readonly attributeTypeMap: Array<{name: string, baseName: string, type: string, format: string}> = [
        {
            "name": "accessKey",
            "baseName": "accessKey",
            "type": "string",
            "format": ""
        },
        {
            "name": "queueDsn",
            "baseName": "queueDsn",
            "type": "URI",
            "format": "uri"
        },
        {
            "name": "region",
            "baseName": "region",
            "type": "string",
            "format": ""
        },
        {
            "name": "secretKey",
            "baseName": "secretKey",
            "type": "string",
            "format": ""
        },
        {
            "name": "type",
            "baseName": "type",
            "type": "SinkInOneOf1TypeEnum",
            "format": ""
        }    ];

    static getAttributeTypeMap() {
        return SinkInOneOf1.attributeTypeMap;
    }
    
    public constructor() {
    }
}


export type SinkInOneOf1TypeEnum = "sqs" ;
