// svix/src/openapi/models/SinkOut.ts
/**
 * Svix API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.1.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { KafkaSecurityProtocolType } from './KafkaSecurityProtocolType';
import { SinkInOneOf } from './SinkInOneOf';
import { SinkInOneOf1 } from './SinkInOneOf1';
import { SinkInOneOf2 } from './SinkInOneOf2';
import { SinkInOneOf3 } from './SinkInOneOf3';
import { URI } from './URI';
import { HttpFile } from '../http/http';

export class SinkOut {
    /**
    * List of message channels this endpoint listens to (omit for all)
    */
    'channels'?: Array<string>;
    'createdAt': Date;
    /**
    * An example endpoint name
    */
    'description': string;
    'disabled'?: boolean;
    'filterTypes'?: Array<string>;
    /**
    * The ep's ID
    */
    'id': string;
    'metadata': { [key: string]: string; };
    'rateLimit'?: number;
    /**
    * Optional unique identifier for the endpoint
    */
    'uid'?: string;
    'updatedAt': Date;
    'routingKey': string;
    'type': SinkOutTypeEnum;
    'uri': string;
    'accessKey': string;
    'queueDsn': URI;
    'region': string;
    'secretKey': string;
    'brokers': string;
    /**
    * Password for SASL, if `security_protocol` is `sasl-ssl`.
    */
    'saslPassword'?: string;
    /**
    * Username for SASL, if `security_protocol` is `sasl-ssl`.
    */
    'saslUsername'?: string;
    'securityProtocol': KafkaSecurityProtocolType;
    'topic': string;
    'url': URI;

    static readonly discriminator: string | undefined = undefined;

    static readonly attributeTypeMap: Array<{name: string, baseName: string, type: string, format: string}> = [
        {
            "name": "channels",
            "baseName": "channels",
            "type": "Array<string>",
            "format": ""
        },
        {
            "name": "createdAt",
            "baseName": "createdAt",
            "type": "Date",
            "format": "date-time"
        },
        {
            "name": "description",
            "baseName": "description",
            "type": "string",
            "format": ""
        },
        {
            "name": "disabled",
            "baseName": "disabled",
            "type": "boolean",
            "format": ""
        },
        {
            "name": "filterTypes",
            "baseName": "filterTypes",
            "type": "Array<string>",
            "format": ""
        },
        {
            "name": "id",
            "baseName": "id",
            "type": "string",
            "format": ""
        },
        {
            "name": "metadata",
            "baseName": "metadata",
            "type": "{ [key: string]: string; }",
            "format": ""
        },
        {
            "name": "rateLimit",
            "baseName": "rateLimit",
            "type": "number",
            "format": "uint16"
        },
        {
            "name": "uid",
            "baseName": "uid",
            "type": "string",
            "format": ""
        },
        {
            "name": "updatedAt",
            "baseName": "updatedAt",
            "type": "Date",
            "format": "date-time"
        },
        {
            "name": "routingKey",
            "baseName": "routingKey",
            "type": "string",
            "format": ""
        },
        {
            "name": "type",
            "baseName": "type",
            "type": "SinkOutTypeEnum",
            "format": ""
        },
        {
            "name": "uri",
            "baseName": "uri",
            "type": "string",
            "format": ""
        },
        {
            "name": "accessKey",
            "baseName": "accessKey",
            "type": "string",
            "format": ""
        },
        {
            "name": "queueDsn",
            "baseName": "queueDsn",
            "type": "URI",
            "format": "uri"
        },
        {
            "name": "region",
            "baseName": "region",
            "type": "string",
            "format": ""
        },
        {
            "name": "secretKey",
            "baseName": "secretKey",
            "type": "string",
            "format": ""
        },
        {
            "name": "brokers",
            "baseName": "brokers",
            "type": "string",
            "format": ""
        },
        {
            "name": "saslPassword",
            "baseName": "saslPassword",
            "type": "string",
            "format": ""
        },
        {
            "name": "saslUsername",
            "baseName": "saslUsername",
            "type": "string",
            "format": ""
        },
        {
            "name": "securityProtocol",
            "baseName": "securityProtocol",
            "type": "KafkaSecurityProtocolType",
            "format": ""
        },
        {
            "name": "topic",
            "baseName": "topic",
            "type": "string",
            "format": ""
        },
        {
            "name": "url",
            "baseName": "url",
            "type": "URI",
            "format": "uri"
        }    ];

    static getAttributeTypeMap() {
        return SinkOut.attributeTypeMap;
    }
    
    public constructor() {
    }
}


export type SinkOutTypeEnum = "http" ;
