// svix/src/openapi/models/StreamSinkPatch.ts
/**
 * Svix API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.1.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RedshiftConfig } from './RedshiftConfig';
import { SinkPayloadFormat } from './SinkPayloadFormat';
import { SinkStatusIn } from './SinkStatusIn';
import { StreamSinkInOneOf } from './StreamSinkInOneOf';
import { StreamSinkInOneOf1 } from './StreamSinkInOneOf1';
import { StreamSinkInOneOf2 } from './StreamSinkInOneOf2';
import { StreamSinkInOneOf3 } from './StreamSinkInOneOf3';
import { StreamSinkInOneOf4 } from './StreamSinkInOneOf4';
import { StreamSinkInOneOf5 } from './StreamSinkInOneOf5';
import { StreamSinkInOneOf6 } from './StreamSinkInOneOf6';
import { HttpFile } from '../http/http';

export class StreamSinkPatch {
    'batchSize'?: number;
    'format'?: SinkPayloadFormat;
    'maxWaitSecs'?: number;
    'status'?: SinkStatusIn;
    'uid'?: string;
    'config': RedshiftConfig;
    'type': StreamSinkPatchTypeEnum;

    static readonly discriminator: string | undefined = undefined;

    static readonly attributeTypeMap: Array<{name: string, baseName: string, type: string, format: string}> = [
        {
            "name": "batchSize",
            "baseName": "batchSize",
            "type": "number",
            "format": "uint16"
        },
        {
            "name": "format",
            "baseName": "format",
            "type": "SinkPayloadFormat",
            "format": ""
        },
        {
            "name": "maxWaitSecs",
            "baseName": "maxWaitSecs",
            "type": "number",
            "format": "uint16"
        },
        {
            "name": "status",
            "baseName": "status",
            "type": "SinkStatusIn",
            "format": ""
        },
        {
            "name": "uid",
            "baseName": "uid",
            "type": "string",
            "format": ""
        },
        {
            "name": "config",
            "baseName": "config",
            "type": "RedshiftConfig",
            "format": ""
        },
        {
            "name": "type",
            "baseName": "type",
            "type": "StreamSinkPatchTypeEnum",
            "format": ""
        }    ];

    static getAttributeTypeMap() {
        return StreamSinkPatch.attributeTypeMap;
    }
    
    public constructor() {
    }
}


export type StreamSinkPatchTypeEnum = "redshift" ;
